New MGXS for chi-prompt, prompt-nu-fission, and velocity

samuelshaner opened the pull request:
This PR adds in a few new MGXS objects that will be of use in generating data for kinetics simulations. Also, a few typos in the python api have been fixed. All MGXS tests have been updated to include the new MGXS types.

paulromano commented:
@samuelshaner Looks like one test fails with OpenMP and/or MPI turned on

paulromano commented on openmc/mgxs/mgxs.py:
```diff
@@ -1935,7 +1948,7 @@ def print_xs(self, subdomains='all', nuclides='all', xs_type='macro'):
 
 class TotalXS(MGXS):
-    r"""A total multi-group cross section.
+    """A total multi-group cross section.
```
All these docstrings need to be raw in order for the equations to render properly in the documentation.

samuelshaner replied:
Ok, so that `r` was there for a reason.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -427,7 +430,11 @@ def get_mgxs(mgxs_type, domain=None, domain_type=None,
         Parameters
         ----------
-        mgxs_type : {'total', 'transport', 'nu-transport', 'absorption', 'capture', 'fission', 'nu-fission', 'kappa-fission', 'scatter', 'nu-scatter', 'scatter matrix', 'nu-scatter matrix', 'multiplicity matrix', 'nu-fission matrix', chi'}
+        mgxs_type : {'total', 'transport', 'nu-transport', 'absorption',
```
This must actually be on one line in order for Sphinx to properly parse it. You're actually the third developer to try this, I wish we could add a comment stating that this must unfortunately remain on a single line for this reason, but this already is a comment!

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+
+class ChiPrompt(Chi):
+    """The prompt fission spectrum.
```
You'll need an `r` here as well so that the Sphinx can parse the equation.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+        return self._xs_tally
+
+    def get_slice(self, nuclides=[], groups=[]):
```
It's too bad some of the specialized methods can't be inherited from `Chi`, but I see the challenge of designing the code to make that possible.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+        Returns
+        -------
+        openmc.mgxs.MGXS
```
Make the return value `openmc.mgxs.ChiPrompt`. I think it is more generally `openmc.mgxs.MGXS` for the general `get_slice(...)` method implemented at the abstract `MGXS` class level.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+            # Slice nu-fission-out tally along energyout filter
+            prompt_nu_fission_out = slice_xs.tallies['nu-fission-out'] - \
+                                    slice_xs.tallies['delayed-nu-fission-out']
+            tally_slice = prompt_nu_fission_out\
```
Can you revise this to make it look a little cleaner:

```python
tally_slice = prompt_nu_fission_out.get_slice(
    filters=['energyout'], filter_bins=filter_bins)
```

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+            slice_xs._tallies['prompt-nu-fission-out'] = tally_slice
+
+        # Add energy filter back to nu-fission-in tallies
+        slice_xs._tallies['prompt-nu-fission-in'].add_filter(energy_filter)
```
You can use the `tallies` property directly here

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+        """
+
+        if not self.can_merge(other):
+            raise ValueError('Unable to merge ChiPrompt')
```
raise a `NotImplementedError` instead

wbinventor replied:
Oh never mind, I see.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+            if nuclides == 'sum' or nuclides == ['sum']:
+
+                # Retrieve the fission production tallies
+                prompt_nu_fission_in = self.tallies['nu-fission-in'] - \
```
Just curious, do you think it is worth storing `prompt_nu_fission_in` and `prompt_nu_fission_out` in the `tallies` attribute when the cross section is first computed? This would avoid redundant calculations in the event that methods like `get_xs(...)` are called a bunch of times (which can be a common use case in my experience).

wbinventor replied:
In fact, I think that if you store each of these tallies indexed by `"nu-fission-in"` and `"nu-fission-out"`, then you can simply inherit a lot of this methods from the parent `Chi` class. Do you see that as well, or am I being overly optimistic?

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+                # Sum out all nuclides
+                nuclides = self.get_all_nuclides()
+                prompt_nu_fission_in = prompt_nu_fission_in.summation\
+                                       (nuclides=nuclides)
```
Put the opening parentheses on the top line:

```python
prompt_nu_fission_in = prompt_nu_fission_in.summation(
    nuclides=nuclides)
```

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+    :attr:`Velocity.tallies` property, which can then be appended to a
+    :class:`openmc.Tallies` instance.
+
+    For post-processing, the :meth:`MGXS.load_from_statepoint` will pull in the
```
Can you specialize this docstring a bit and state that this is a flux-weighted average velocity in each group? If that is indeed the case of course. Also, what are the units?

wbinventor replied:
I understand that it is obvious that it is flux-weighted from the equation below, but might be nice to explicitly state it in text as well.b

samuelshaner replied:
Sure, I'll specialize the docstring. The velocity is actually computed by first computing the flux weighted inverse velocity and then taking the inverse of that so more detail would probably be good to clarify that.

The units of velocity are cm/s. I'll note that here. While m/s seems like a more common unit for velocity, I chose to keep it as cm/s since the flux and cross sections all use cm.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+        return self._xs_tally
+
+    def print_xs(self, subdomains='all', nuclides='all', xs_type='macro'):
```
Why can't this method be inherited from `MGXS`?

samuelshaner replied:
It could, but the units would be listed as cm^-1, which might be confusing. Other than changing line 5214, it is the same as MGXS.print_xs(). One option is to have a `units` attribute for the MGXS class that is used in constructing the header. I could see other MGXS being created down the road that would have other units so this might be useful in more than just the velocity. Any objections to adding a units attribute? If not, I'll go ahead and add it.

wbinventor commented:
@samuelshaner this is looking pretty good! In addition to my inline comments, take a look through the "MGXS" Jupyter Notebooks and update them with your new `MGXS` subclasses. At least one of them includes a bulleted list of all available subclasses.

nelsonag commented on openmc/mgxs/mgxs.py:
```diff
@@ -4615,3 +4628,728 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+    @property
+    def scores(self):
+        return ['delayed-nu-fission', 'delayed-nu-fission',
```
Seems like a lot of this code could be simplified if OpenMC had a prompt-nu-fission score.  What do you think?

samuelshaner replied:
Yeah, it would be simplified a lot. We have a `delayed-nu-fission` score, so adding a `prompt-nu-fission` score seems logical. I'll look into implementing this score and simplifying the ChiPrompt class.

nelsonag commented on src/tally.F90:
```diff
@@ -441,6 +441,69 @@ subroutine score_general_ce(p, t, start_index, filter_index, i_nuclide, &
         end if
 
 
+      case (SCORE_PROMPT_NU_FISSION)
```
On its own this looks fine.  But it does make me wonder if this can be combined with the delayed version.  It looks like the differences would be in the EMISSION_PROMPT constant and the call to score_fission_prompt_eout.  I think socre_fission_prompt_eout can be generalized to take an argument for the delayed group or prompt (i.e., 0 is prompt, a number > 0 is the group).  What do you think?

samuelshaner replied:
I think it could be simplified such that the total/prompt/delayed are all handled in the same method via an integer argument to the routine. I'll look into this.

samuelshaner replied:
I just pushed a modification such that there is only one routine, `score_fission_eout_ce`, which accepts the `score_bin` as an argument and scores the emitted neutrons if their emission (prompt or delayed) agrees with the requested emission type.

paulromano replied:
I like the simplification!

samuelshaner commented:
I keep having a problem where the `test_mgxs_library_hdf5` test will pass for hdf5_normal, but fail for all other configurations. I have tried reinstalling MPI and HDF5 (which was painful) and running on different systems, but keep getting similar results. Does anyone have any guidance on what might be going on?

samuelshaner commented on openmc/mgxs/mgxs.py:
```diff
@@ -10,6 +10,11 @@
+# Require numpy to print output in scientific notation to 6 decimal places.
+# This is needed to avoid round off error when large numbers are printed,
+# which can cause tests to fail for different build configurations.
+np.set_printoptions(formatter={'float': lambda x: format(x, '8.6E')})
```
When the velocities were printed to file in the `test_mgxs_library_hdf5`, they were printed to ~16 digits of precision which caused different build configurations to disagree on the last 2-3 digits. This line requires numpy to always print out numbers in scientific notation to six decimal places. I'll be the first to admit this is kind of a hack. Nevertheless it is an issue that will come back if we just make a special case for printing the velocities where the values are truncated after a certain number of decimal places. @paulromano @wbinventor and others - if you have any thoughts on a more elegant (and preferably general) solution, I'm all ears.

wbinventor replied:
I'm curious whether this option can be set in our test suite *rather* than setting it universally for the API? Perhaps it can be set in the ["testing_harness.py"](https://github.com/mit-crpg/openmc/blob/develop/tests/testing_harness.py) file?

samuelshaner replied:
That seems feasible. I'll put it there instead.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4632,470 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+        self._rxn_type = 'chi-prompt'
+
+    @property
+    def scores(self):
```
This looks beautiful!

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4632,470 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+    def scores(self):
+        return ['prompt-nu-fission', 'prompt-nu-fission']
+
+    def merge(self, other):
```
Will the parent class `merge` method not work? If not, make note of why it won't work in a comment somewhere. Seems like it should with this new class implementation.

samuelshaner replied:
The parent class merge works just fine. I inherited the method because the error message refers to the class. This method does the error checking and then calls the parent class merge. I'll just use the parent method and make the error message a bit ambiguous about whether Chi, ChiPrompt, or ChiDelayed is being looked at.

wbinventor replied:
sounds good

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4632,470 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+    """
+
+    def __init__(self, domain=None, domain_type=None,
```
This also looks great.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4632,470 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+                    string += '{0: <16}=\t{1}\n'.format('\tNuclide', nuclide)
+
+                # Build header for cross section type
+                string += '{0: <16}\n'.format\
```
The `Velocity.print_xs(...)` method is exactly the same as the `MGXS.print_xs(...)` method save for this block for units, correct? I like your idea of creating a `MGXS.units` method which would allow us to eliminate this redundant code. Perhaps `units` can take an `xs_type` parameter with accepted `'macro'` or `'micro'` values. The `MGXS.units` method would return the strings `'cm^-1'` or `'barns'` while `Velocity.units` method would simply return `'cm/second'` in each case. What do you think?

samuelshaner replied:
That sounds great. So would units be a dictionary with `micro` and `macro` keys or should there just be a `get_units(xs_type='micro')` method that accepts an `xs_type` argument?

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4632,470 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+    def __init__(self, domain=None, domain_type=None,
+                 groups=None, by_nuclide=False, name=''):
+        super(PromptNuFissionXS, self).__init__(domain, domain_type, groups,
```
This also looks fantastic.

wbinventor commented on tests/test_mgxs_library_hdf5/results_true.dat:
```diff
@@ -1,195 +1,222 @@
 domain=10000 type=total
-[ 0.41482549  0.66016992]
```
These results only changed due to the new NumPy print options setting right?

samuelshaner replied:
correct

wbinventor commented on tests/test_mgxs_library_no_nuclides/results_true.dat:
```diff
@@ -75,6 +75,15 @@
 1     10000          1   total   1.0   0.046071
 0     10000          2   total   0.0   0.000000
+   material  group out nuclide      mean  std. dev.
+1     10000          1   total  0.991442   0.050935
```
If I understand correctly, these results indicate that `ChiPrompt` is not properly normalized. Or does this mean that 0.991442 neutrons emitted from fission are prompt and in the fast group?

wbinventor commented on tests/test_mgxs_library_no_nuclides/results_true.dat:
```diff
@@ -152,6 +161,15 @@
 1     10001          1   total   0.0        0.0
 0     10001          2   total   0.0        0.0
+   material  group out nuclide  mean  std. dev.
+1     10001          1   total   0.0        0.0
```
Should `ChiPrompt` be zero in each bin here?

samuelshaner replied:
Yeah. Chi is right above it and it is zero so ChiPrompt should be zero as well.

wbinventor commented:
This PR looks pretty good for me save for my questions and comment about the `Velocity.print_xs(...)` method.

samuelshaner commented:
I think this PR is ready to be merged or further reviewed. In `testing_harness.py`, I added a requirement that numpy and pandas print out arrays in scientific notation to avoid any slight differences in large values (e.g. velocities) that result because different configurations add numbers in different orders. Additionally, I added a `prompt-nu-fission` score and simplified the way nu-fission, prompt-nu-fission, and delayed-nu-fission are tallied when an energy out filter is present in `tally.F90`.

wbinventor commented:
This looks fine to me, but I'd like a more knowledgeable Fortran developer to take a look at that section of the diff before merging this in.

paulromano commented on openmc/mgxs/mgxs.py:
```diff
@@ -1530,6 +1538,31 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+        if xs_type == 'macro':
+            return 'cm^-1'
+        else:
+            return 'barns'
```
`return 'cm^-1' if xs_type == 'macro' else 'barns'` is more concise

paulromano commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4649,415 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+        if xs_type == 'macro':
+            return 'cm/second'
+        else:
+            raise ValueError('Unable to return the units of Velocity for ' +
```
The `+` at the end is superfluous (Python concatenates whitespace-separated strings).

paulromano commented on src/tally.F90:
```diff
@@ -1555,13 +1626,51 @@ subroutine score_fission_eout_ce(p, t, i_score)
-      i_filter = sum((matching_bins(1:t%n_filters) - 1) * t % stride) + 1
+      ! Case for tallying prompt neutrons
+      if (score_bin == SCORE_NU_FISSION .OR. \
+        (score_bin == SCORE_PROMPT_NU_FISSION .AND. g == 0)) then
```
Lowercase `.or.` and `.and.`

paulromano commented on src/tally.F90:
```diff
@@ -1555,13 +1626,51 @@ subroutine score_fission_eout_ce(p, t, i_score)
+             t % results(i_score, i_filter) % value + score
+
+      ! Case for tallying delayed emissions
+      else if (score_bin == SCORE_DELAYED_NU_FISSION .AND. g /= 0) then
```
lowercase `.and.`

paulromano commented on src/tally.F90:
```diff
@@ -1555,13 +1626,51 @@ subroutine score_fission_eout_ce(p, t, i_score)
-      ! determine scoring index
-      i_filter = sum((matching_bins(1:t%n_filters) - 1) * t % stride) + 1
+      ! Case for tallying prompt neutrons
+      if (score_bin == SCORE_NU_FISSION .OR. \
```
The backslash should be an ampersand. Also, continuation lines [should be indented by 5 spaces](http://openmc.readthedocs.io/en/latest/devguide/styleguide.html#indentation).

paulromano commented on src/tally.F90:
```diff
@@ -441,6 +441,67 @@ subroutine score_general_ce(p, t, start_index, filter_index, i_nuclide, &
+            ! score the number of particles that were banked in the fission
+            ! bank as prompt neutrons. Since this was weighted by 1/keff, we
+            ! multiply by keff to get the proper score.
+            score = keff * p % wgt_bank * (1.0 - sum(p % n_delayed_bank) &
```
Use the constant `ONE` instead of 1.0

paulromano commented on src/tally.F90:
```diff
@@ -441,6 +441,67 @@ subroutine score_general_ce(p, t, start_index, filter_index, i_nuclide, &
+            ! bank as prompt neutrons. Since this was weighted by 1/keff, we
+            ! multiply by keff to get the proper score.
+            score = keff * p % wgt_bank * (1.0 - sum(p % n_delayed_bank) &
+                 / real(p % n_bank))
```
`real(p % n_bank, 8)` is safer. Just worried about losing precision here converting integers to reals

paulromano commented on src/tally.F90:
```diff
@@ -1555,13 +1626,51 @@ subroutine score_fission_eout_ce(p, t, i_score)
+        i_filter = sum((matching_bins(1:t%n_filters) - 1) * t % stride) + 1
+
+        ! Add score to tally
+        !$omp atomic
```
The prevailing style is to have OpenMP directives not indented at all.

paulromano commented on src/tally.F90:
```diff
@@ -1555,13 +1626,51 @@ subroutine score_fission_eout_ce(p, t, i_score)
+            ! delayed group of this bin
+            if (d == g) then
+              call score_fission_delayed_dg(t, d_bin, score, i_score)
+            end if
```
I'd prefer the one-liner `if (d == g) call score_fission_delayed_dg(t, d_bin, score, i_score)`

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -1530,6 +1538,31 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
         return df
 
+    def get_units(self, xs_type='macro'):
+        r"""Returns the units of a MGXS.
```
Neglected to mention this yesterday, but the "r" isn't necessary to format this docstring since it doesn't include any math (it will still work, but the convention elsewhere has been to format the strings only when necessary). Also, the comment can be shortened to simply "This method returns the units of a MGXS based on a desired xs_type."

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4649,415 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
         return df
+
+    def get_units(self, xs_type='macro'):
+        r"""Returns the units of Chi.
```
Remove the "r"

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4649,415 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+        return self._xs_tally
+
+    def get_units(self, xs_type='macro'):
+        r"""Returns the units of Velocity.
```
Remove the "r"

paulromano commented:
@samuelshaner The differences in test results that you overcame by forcing numpy and pandas to only display 7 significant digits worry me slightly. We should be able to match results to greater than single precision. Is there a reason we don't go for more, say, 12 digits?

paulromano commented on tests/testing_harness.py:
```diff
@@ -9,6 +9,13 @@
+# places. This is needed to avoid round off error when large numbers are
+# printed, which can cause tests to fail for different build configurations.
+np.set_printoptions(formatter={'float': lambda x: format(x, '14.12E')})
+pd.set_option('display.float_format', lambda x: '%14.12E' % x)
```
The following is a little cleaner, and IMO more Pythonic:
```Python
np.set_printoptions(formatter={'float': '{:.12e}'.format})
pd.options.display.float_format = '{:.12e}'.format
```

paulromano commented:
@samuelshaner Are you ok if we target this for v0.9 instead of the 0.8?

samuelshaner commented:
@paulromano - the root of the issue is that some tallies produce small values (on the order of 1e-12 for some tests) while other tallies produce large values (on the order of 1e7). The small tallies match between different configurations down to close to machine epsilon for double precision, but large tallies (e.g. the Velocity MGXS tally) only match down to 7 or 8 decimal places. The Velocity MGXS tally divides a number near unity (the flux tally) by a small number (the flux weighted inverse velocity). The uncertainty due to machine precision in the flux weighted inverse velocity tally will cause the resultant velocity to have an uncertainty (due to machine precision) much larger than machine epsilon. My sense is that using the default of 8 digits of precision for Numpy and Pandas puts us below the expected uncertainty due to machine precision and thus we would expect different results if the values in the tallies were added in a different order, which could be the case for different configurations (I believe). I see two options:
- **Output all the numbers in scientific notation**: This would allow us to compare a certain number of significant figures in the output values, but we are limited by small values only having a few significant figures until we hit machine epsilon.
- **Output all the numbers to a set precision**: This is limited by the machine precision of values like the MGXS Velocity tally, which is much greater than machine epsilon. Setting a relatively high precision limit (e.g. 1e-8) would cause many small values to be zero.  

I think the scientific notation approach with 7 significant figures (the most we can have before tests start failing) is preferred, but am open to other options as there are clear issues with both approaches. In short, I don't think we can go beyond ~7 significant figures or ~7 digits of precision without having tests start to fail due to round off error.

Targeting v0.9 instead of v0.8 is fine with me.

samuelshaner commented:
Given the difficulty with getting the test suite to produce the same results with different configurations when I added the Velocity MGXS, I've decided to change the Velocity MGXS to InverseVelocity. With this change, the tests now pass for all configurations without modifying the way the test suite prints python arrays. Kinetics calculations require the inverse velocity, so I don't see this change as requiring any extra steps in downstream calculations (at least for what I am doing).

If this is an agreeable switch, this PR should be ready for further review or merging.

paulromano commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4644,392 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+    .. math::
+
+       \langle \nu\sigma_{f,\rightarrow g}^p \phi \rangle &= \int_{r \in V} dr
```
Did you really want the comma in `f,\rightarrow g`?

wbinventor replied:
Shouldn't it be `f, g'\rightarrow g`?

paulromano replied:
Right, that's what I really meant...

samuelshaner replied:
Yep, I just copied and pasted from `Chi` so I'll change it there as well.

paulromano commented on openmc/mgxs/mgxs.py:
```diff
@@ -4618,3 +4644,392 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
+
+       \langle \nu\sigma_{f,\rightarrow g}^p \phi \rangle &= \int_{r \in V} dr
+       \int_{4\pi} d\Omega' \int_0^\infty dE' \int_{E_g}^{E_{g-1}} dE \; \chi(E)
+       \nu\sigma_f^p (r, E') \psi(r, E', \Omega')\\
```
How do you feel about having `\nu_p \sigma_f` instead?

wbinventor replied:
I prefer this as well.

samuelshaner replied:
I think that more accurately reflects what the code actually does. I'll make the change.

paulromano commented:
@samuelshaner This is good to go once those small comments are resolved. Thanks for the work on this, and sorry you wasted an unknown number of hours of your life trying to get the test results matching!

wbinventor commented:
Looks good to me as well! Thanks for working to make the new features rock solid!

samuelshaner commented:
Thanks for the quick review @paulromano and @wbinventor. I just addressed your comments and its passing all tests so should be ready to merge.

wbinventor commented:
One thought from this PR - moving forward, it would be nice to have an efficient way to deploy code on Travis without having to add a commit each time (re: there were a lot of garbage commits in this PR for this purpose). Of course one could create lots of test branches / commits in order to figure out a working solution to test failures, or cherry pick out the relevant commits. @paulromano do you have any thoughts on how best to avoid this in the future?

paulromano commented:
@wbinventor I don't have any great suggestions. I always try to set up my system so that I can reproduce Travis failures and then fix them locally before just pushing something to see if it works. But, Travis and/or gfortran 4.6 is very finicky sometimes. If the history is already messy when the pull request is first made, I'd say rebase, squash, and clean up the history to whatever extent possible beforehand. I also personally don't mind changing history in a pull request via force-pushes, but I gather not everyone likes to live that dangerously.

smharper commented:
@wbinventor, I assume you're talking about tests that work locally but do not fail on Travis?  The way I handle it is by opening a new branch in my repo, pushing to that branch, and having Travis check that.  Once you sign up for it, Travis will automatically run any commits you push to your repo.  The check for pushes is less rigorous than the PR tests (it misses things like OpenMP) but it catches most errors.

wbinventor commented:
@smharper that is exactly what I do now as well. If that seems to be the best approach, then perhaps we should ask that developers take that approach in the future to avoid unnecessary commits which take a trial and error approach to correct Travis failures.
